### PR TITLE
Implement Fuzzing Matching leveraging CPE Regular Expressions

### PR DESCRIPTION
**Pull request by officerNordberg**
-also lucene fuzzing ~
@stevespringett had to abandon the previous PR

Signed-off-by: Drew Thompson <drew.michael.thompson+ghdco@gmail.com>

### CODE REVIEW

**Comment by msymons** [comment]
Replaces PR #1486 (useful thoughts and commentary included there).

**Comment by officerNordberg** [comment]
@msymons @nscuro Has anyone been reviewing this PR?  I keep having to resolve conflicts while it sits in limbo.

**Comment by stevespringett** [comment]
Ok good. Wasn't aware you were complete with this. @nscuro can we target this for 4.6?

File: src/test/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManangerTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import us.springett.parsers.cpe.values.Part;
+
+import java.util.List;
+import java.util.regex.Pattern;
+
+import static org.mockito.Mockito.*;
+
+import static org.junit.Assert.*;
+
+public class FuzzyVulnerableSoftwareSearchManangerTest {
```
Comment (nscuro) [inline src/test/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManangerTest.java:23]: It'd be great if we could have non-`@Ignore`d tests for this. Off the top of my head I don't know how we can construct and reset the global Lucene index, but there should be a way to do it. Maybe it's already done in some other test?

File: src/test/resources/logback-test.xml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (nscuro) [inline src/test/resources/logback-test.xml:1]: Please remove this file. I see why it's beneficial to have when running tests locally, but when running the entire test suite in CI it's flooding the build logs.

File: src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    private static final Logger LOGGER = Logger.getLogger(FuzzyVulnerableSoftwareSearchMananger.class);
+    private static final Set<String> DO_NOT_FUZZ = Collections.unmodifiableSet(Set.of("util", "utils", "url", "xml"));
+
+    private final boolean excludeComponentsWithPurl;
+    private final Set<String> SKIP_LUCENE_FUZZING_FOR_TYPE = Collections.unmodifiableSet(Sets.newHashSet("golang"));
+    public FuzzyVulnerableSoftwareSearchMananger(boolean excludeComponentsWithPurl) {
+        this.excludeComponentsWithPurl = excludeComponentsWithPurl;
+    }
+
+    private class SearchTerm {
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java:40]: This class should be static.

```suggestion
    private static class SearchTerm {
```

**Review by nscuro — CHANGES_REQUESTED** [review]
Sorry for the long wait @officerNordberg, I already had some pending comments but always got distracted with something else, and forgot to submit them.

This would be nice to have in 4.6.

File: src/test/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManangerTest.java (hunk 1 above)
Comment (officerNordberg) [reply src/test/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManangerTest.java:23]: I really don't know how to write the integration tests used by this project, I tried mocking stuff but had difficulty due to how Singletons are used.

File: src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+                String nameToFuzz = component.getName();
+                if (parsedCpe != null) {
+                    part = parsedCpe.getPart();
+                    vendor = parsedCpe.getVendor();
+                    searches.add(new SearchTerm(parsedCpe.getVendor(), parsedCpe.getProduct()));
+                    nameToFuzz = parsedCpe.getProduct();
+                }
+                if (component.getPurl() != null) {
+                    if (component.getPurl().getType().equals("golang")) {
+                        String namespace = Arrays.stream(component.getPurl().getNamespace().split("/")).reduce((first, second) -> second)
```
[the remark this answers is not in the document]
Comment (officerNordberg) [reply src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java]: Ugh, I know why I went down this path but it's also working as expected currently so I'm not sure if I lost my mind or fixed the bom.

**Comment by officerNordberg** [comment]
@nscuro looks like I need to abandon this PR yet again.  The "helpful" github feature where you suggest code and I commit it fails DCO and I can't rebase because I've been pulling in master.

File: src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        final SearchResult searchResult = new SearchResult();
+        final List<Map<String, String>> resultSet = new ArrayList<>();
+        IndexManager indexManager = VulnerableSoftwareIndexer.getInstance();
+        try {
+            final Query query = indexManager.getQueryParser().parse(luceneQuery);
+            final TopDocs results = indexManager.getIndexSearcher().search(query,1000);
+
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Searching for: " + luceneQuery + " - Total Hits: " + results.totalHits);
+            }
```
[the remark this answers is not in the document]
Comment (officerNordberg) [reply src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchMananger.java]: actually, my bad maybe alpine has added that or I was wrong all along.  Wish I could just use known frameworks

**Pull request closed without merging**